### ISSUE TITLE
4.0 doesn't support require in nodeJS

### ISSUE THREAD

**Issue by bitdom8** [ai/nanoid#364]
const { customAlphabet } = require('nanoid')
                           ^

Error [ERR_REQUIRE_ESM]: require() of ES Module C:\Users\Administrator\Documents\serverws\node_modules\.pnpm\nanoid@4.0.0\node_modules\nanoid\index.js from C:\Users\Administrator\Documents\serverws\routes.js not supported.

**Comment by ai** [comment]
Yes. It is indented. [4.0](https://github.com/ai/nanoid/blob/main/CHANGELOG.md) removed CommonJS support (`require()`).

You can move your app to ESM a use `import` instead of `require`.

Or you can stack with 3.x branch. We will support 3.x for a year.

**Comment by bitdom8** [comment]
Thanks you sir @ai

**Comment by shellscape** [comment]
Yes, we'll continue to use the 3.x version branch as long as we're able to. And still grateful for that. 

> Or you can stack with 3.x branch. We will support 3.x for a year.

After a year, 3.x will no longer have support, and that's sad. Businesses will not be updated to ESM-only dependencies and codebases in that time. My best-guess estimate, based on watching the ecosystem and commercial adoption since v0.10, is that it will take the ecosystem at least 5 more years to get there. I'm not suggesting you support 3.x for 5 years, only providing my take on adoption of ESM-only.

**Comment by ai** [comment]
> We will support 3.x for a year.

It is not an exactly a year, but a year minimum.

Anyway I need Nano ID 3 in PostCSS.

**Comment by ljharb** [comment]
@ai can you elaborate on *why* this decision was made?

It's worth noting that every single module that's gone ESM-only has gotten virtually no adoption for those versions; you can check on their npm package page under the "versions" tab. nano, so far, has only 1.57% after 5 months; there are packages that have been made this change for almost 2 years with similarly low uptake.

**Comment by ljharb** [comment]
@ai sure, that's fine. I'd still love to understand *why* you made this change.

**Comment by ai** [comment]
Sure, my personal reasons:

1. I see “nano” projects as something small and very effective. Having duplicates for every file (for ESM and CJS versions) in npm package doesn’t satisfy this goal.
2.  Now I have only 2 types of projects: pure CJS (like `postcss`) and pure ESM (like most of `nano*` projects). Before 4.x I have 3 types and must to always remember using special tools for dual ESM/CJS packages.
3. All my active project was moved to ESM (PostCSS and Autoprefixer are on maintenance). For me, it is OK to keep using 3.x on the projects on maintenance stage.
4. I stop believing in open source community to be really extra care of every use case and invest my time there.
5. Supporting 2 branches is for me a good compromise between “keeping project simple”, “simplify maintenance” and still have a solution for every user’s case. The only problem is esthetic problem of some users (who want to use only latest versions, but didn’t migrate to ESM), but I do not want to invest my time to it because of previous step.

**Comment by ljharb** [comment]
That all makes sense - but what’s the value of “pure esm” given that esm can import CJS?

**Comment by ai** [comment]
@ljharb originally we added ESM to use package from ESM-compatible CDNs and have destructor in import API `import {  customAlphabet } from 'nanoid'`.

I am not sure that migration to CJS will keep project compatible with all edge cases. But maybe I am wrong.

**Comment by ljharb** [comment]
ESM can named-import from CJS if the CJS is babel-transpiled.

npm packages simply aren't (in any consistent sense) usable from a CDN without a build process; any such build process can easily handle CJS modules just as well as ESM modules. I agree that you can't get a no-build-process native ESM setup working with a CJS module, but since many, many 9's of npm packages are in CJS, that's not really a viable option anyways.

**Comment by ai** [comment]
> ESM can named-import from CJS if the CJS is babel-transpiled.

1. What does it mean?
2. Where exactly it will work with 100% guarantee? I need to support Node.js, Parcel 1 & 2, webpack, Rollup, esbuild, Metro.

> npm packages simply aren't (in any consistent sense) usable from a CDN without a build process

Nano ID not only can but even specially created to be served from the simplest CDN:

```js
// Not sure that it works because of GitHub content-type
import { nanoid } from 'https://raw.githubusercontent.com/ai/nanoid/main/nanoid.js'
```

> any such build process can easily handle CJS modules just as well as ESM modules

Are we 100% sure with it, and about what CDNs are we sure?

A lot of people were forced me to release CJS/ESM version (2.x→3.x) because they told that their ESM CDN supports only ESM packages.

**Comment by ljharb** [comment]
> What does it mean?

node's ESM implementation uses https://npmjs.com/cjs-module-lexer to try to infer named imports from CJS files. It works especially well for transpiler output.

> Where exactly it will work with 100% guarantee?

By definition, "what node does" is the correct thing, so if any of those bundlers don't support this, then they're broken :-) That said, I'm pretty sure all of these support this, or they wouldn't be able to support the majority of the ecosystem's existing practices.

> Are we 100% sure with it, and about what CDNs are we sure?

I'm 100% sure it *could* be done. I certainly haven't audited existing CDNs to figure out what capabilities they have (mostly because using CDNs isn't a best practice anymore - it's best to control the serving of your own assets).

There are only a very, very small number of authors that have chosen to publish ESM-only packages. If ESM CDNs want to gain adoption, they have no choice but to support CJS eventually anyways - why make life harder for your existing consumers in the meantime by switching to ESM-only? "ESM CDN" is a much less important use case than "every existing node user who uses your package", I'd expect.

**Comment by shellscape** [comment]
For what it's worth, here's everything Sindre has compiled on his move to ESM for all of his stuff: https://github.com/sindresorhus/meta/discussions/15. @wooorm as done the same for rehype/remark/unified and supporting packages as well. I don't like it and don't agree with their reasoning, but it's more than a handful of packages affected.

**Comment by ljharb** [comment]
@shellscape indeed, but it's still only a handful of *authors*

**Comment by ai** [comment]
> By definition, "what node does" is the correct thing, so if any of those bundlers don't support this, then they're broken :-)

@ljharb :D. On one hand I have “3.x ESM+CJS & 4.x ESM” solution which works perfectly and has a single problem of not be so beautiful for CJS cases.

On another hand, I have months of my life spending fixing many edge cases (which we already fixed for 3.x) for rare bundlers and ESM CDNs.

I think you see the reason of why I prefer to keep it. Update to 4.x if you want to use latest versions. CJS is outdated too.

**Comment by ljharb** [comment]
CJS isn’t outdated; it’s a first-class module system for node, and will be supported for the foreseeable future - ESM will one day reach feature parity with CJS, although it’s not there yet.

I’m sorry you had to spend so much time on v4; i maintain hundreds of packages that are just CJS and haven’t had to spend any time on supporting ESM for them ¯\\\_(ツ)_/¯

**Comment by wooorm** [comment]
Fun! Another thread! 😅

> It's worth noting that every single module that's gone ESM-only has gotten virtually no adoption for those versions; you can check on their npm package page under the "versions" tab. nano, so far, has only 1.57% after 5 months; there are packages that have been made this change for almost 2 years with similarly low uptake.
> — @ljharb

This is semver and the npm ecosystem working as designed.
There is no reason why people **must** update *in droves* to a new major version in 5 months.
Slower update cycles **are fine**.
And download counts are terrible!

That’s like saying: hey, I made this new package, I just released it on npm last week, why doesn’t it have millions of downloads yet?!

> what’s the value of “pure esm” given that esm can import CJS?
> — @ljharb

That is true in Node.js but not in other places.

> it's still only a handful of authors
> — @ljharb

It’s… like… all the still active maintainers of lots of popular packages in the JavaScript world that don’t only care about Node.js.

> CJS isn’t outdated
> — @ljharb

I’d argue it very much is. At least for folks that like JavaScript and not necessarily one particular runtime. And if not: it should be!

**Comment by shellscape** [comment]
> I’m sorry you had to spend so much time on v4; i maintain hundreds of packages that are just CJS and haven’t had to spend any time on supporting ESM for them ¯\_(ツ)_/¯

This just isn't constructive and unfortunately on-brand. Just zero need for that.

I don't agree with the decision to move to ESM-only, but I _respect_ @ai and the decision he's made, especially so given the immense value he's contributed to the Node over the years. I credit him for maintaining the CJS branch in the meantime. I understand the arguments from him, Sindre and others, but I don't have to agree with them. They've all explained themselves thoroughly so we can gain perspective on their positions. At this point, that should be enough.

**Comment by ljharb** [comment]
@ai thanks for your explanation. I suspect you'll find that maintaining v3 and v4 is harder than shipping a dual package, and not maintaining v3 will mean folks have to find another solution but that'll be yours to discover :-)

**Comment by nicopowa** [comment]
here's how i fixed v4 with require.
be careful not to call the method too soon because the import is asynchronous.


id.js
```
let id = null;

(async () => {

	const nanoid = await import("nanoid");
	const CustomAlphabet = nanoid.customAlphabet;
	const alphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
	
	id = CustomAlphabet(alphabet, 12);

})();

module.exports = () => id();
```

```
const nanoid = require("./id.js");
//nanoid(); // crash, async import takes a few ms
setTimeout(() => console.log(nanoid()), 100); // working
```

**Comment by shellscape** [comment]
[openbase.com](openbase.com) now displays the ES Module support level (e.g. type of exports) on all of their package pages. If you're continually caught off guard by installing a dependency only to find out it's ESM-only, check the package out there first.

<img width="878" alt="Screen Shot 2022-11-18 at 2 03 41 PM" src="https://user-images.githubusercontent.com/76371/202785490-73387fb2-ec84-46f9-b11e-4ee3fabbd797.png">

<img width="874" alt="Screen Shot 2022-11-18 at 2 03 53 PM" src="https://user-images.githubusercontent.com/76371/202785491-bb0dcdfa-3c06-46f5-b229-bc492e067e8f.png">

<img width="869" alt="Screen Shot 2022-11-18 at 2 04 07 PM" src="https://user-images.githubusercontent.com/76371/202785492-28b540cd-33fc-4da7-b7b2-e5df7b704944.png">

**Comment by ljharb** [comment]
@shellscape since CJS also works in ESM, the only way something should show "CJS" is if it has a `.node` module.

**Comment by shellscape** [comment]
![](https://media.tenor.com/a6jO71lkMmkAAAAC/ronburgundy-agree.gif)

**Comment by ljharb** [comment]
lol no need to block me just for disagreeing with you ¯\\\_(ツ)_/¯

**Comment by kehiy** [comment]
thats worked for me:
`npm uninstall nanoid`
`npm install nanoid@3.3.4 `

**Comment by glassdimlygr** [comment]
@ai has been developing this project for a long time. If you want to assist in shipping a 4x branch that supports CJS, then perhaps you can step forward to help code it. Thank you for your work on this project, @ai.

**Comment by shellscape** [comment]
@ai doesn't need someone running to his defense. and we don't need another useless notification (I realize the irony in saying that whilst writing this comment which will generate a useless notification). Please virtue signal elsewhere.

> Thank you for your work on this project, @ai.

yup. but next time, take it to twitter https://twitter.com/sitnikcode or sponsor him instead of leaving a comment on a closed issue.

**Comment by nicopowa** [comment]
— sorry useless reply —
read and/or copy-paste my monkey patch above or use something else.
